README.md
# [gentleman](https://github.com/h2non/gentleman)-consul [![Build Status](https://travis-ci.org/h2non/gentleman.png)](https://travis-ci.org/h2non/gentleman-consul) [![GoDoc](https://godoc.org/github.com/h2non/gentleman-consul?status.svg)](https://godoc.org/github.com/h2non/gentleman-consul) [![Coverage Status](https://coveralls.io/repos/github/h2non/gentleman-consul/badge.svg?branch=master)](https://coveralls.io/github/h2non/gentleman-consul?branch=master) [![Go Report Card](https://goreportcard.com/badge/github.com/h2non/gentleman-consul)](https://goreportcard.com/report/github.com/h2non/gentleman-consul)

[gentleman](https://github.com/h2non/gentleman)'s v2 plugin for easy service discovery using [Consul](https://www.consul.io).

Provides transparent retry/backoff support for resilient and [reactive](http://www.reactivemanifesto.org) HTTP client capabilities.  
It also allows you to use custom [retry strategies](#custom-retry-strategy), such as [constant](https://godoc.org/github.com/eapache/go-resiliency/retrier#ConstantBackoff) or [exponential](https://godoc.org/github.com/eapache/go-resiliency/retrier#ExponentialBackoff) retries.

## Installation

```bash
go get -u gopkg.in/h2non/gentleman-consul.v2
```

## Versions

- **[v1](https://github.com/h2non/gentleman-consul/tree/v1)** - First version, uses `gentleman@v1`.
- **[v2](https://github.com/h2non/gentleman-consul/tree/master)** - Latest version, uses `gentleman@v2`.

## API

See [godoc reference](https://godoc.org/github.com/h2non/gentleman-consul) for detailed API documentation.

## Examples

See [examples](https://github.com/h2non/gentleman-consul/blob/master/_examples) directory for featured examples.

#### Simple request

```go
package main

import (
  "fmt"

  "gopkg.in/h2non/gentleman.v2"
	"gopkg.in/h2non/gentleman-consul.v2"
)

func main() {
  // Create a new client
  cli := gentleman.New()

  // Register Consul's plugin at client level
  cli.Use(consul.New(consul.NewConfig("demo.consul.io", "web")))

  // Create a new request based on the current client
  req := cli.Request()

  // Set a new header field
  req.SetHeader("Client", "gentleman")

  // Perform the request
  res, err := req.Send()
  if err != nil {
    fmt.Printf("Request error: %s\n", err)
    return
  }
  if !res.Ok {
    fmt.Printf("Invalid server response: %d\n", res.StatusCode)
    return
  }

  // Print response info
  fmt.Printf("Server URL: %s\n", res.RawRequest.URL.String())
  fmt.Printf("Response status: %d\n", res.StatusCode)
  fmt.Printf("Server header: %s\n", res.Header.Get("Server"))
}
```

#### Custom retry strategy

```go
package main

import (
  "fmt"
	"time"

	"gopkg.in/h2non/gentleman.v2"
	"gopkg.in/h2non/gentleman-consul.v1"
	"gopkg.in/eapache/go-resiliency.v1/retrier"
)

func main() {
  // Create a new client
  cli := gentleman.New()

  // Configure Consul plugin
  config := consul.NewConfig("demo.consul.io", "web")

  // Use a custom retrier strategy with max 10 retry attempts
  config.Retrier = retrier.New(retrier.ConstantBackoff(10, time.Duration(25*time.Millisecond)), nil)

  // Register Consul's plugin at client level
  cli.Use(consul.New(config))

  // Create a new request based on the current client
  req := cli.Request()

  // Set a new header field
  req.SetHeader("Client", "gentleman")

  // Perform the request
  res, err := req.Send()
  if err != nil {
    fmt.Printf("Request error: %s\n", err)
    return
  }
  if !res.Ok {
    fmt.Printf("Invalid server response: %d\n", res.StatusCode)
    return
  }

  // Print response info
  fmt.Printf("Server URL: %s\n", res.RawRequest.URL.String())
  fmt.Printf("Response status: %d\n", res.StatusCode)
  fmt.Printf("Server header: %s\n", res.Header.Get("Server"))
}
```

## License

MIT - Tomas Aparicio


History.md

## v2.0.1 / 2017-09-11

  * fix(import): use gentleman-retry.v2
  * fix(style): formatting issues
  * fix(travis): use compatible Go versions

## v2.0.0 / 2017-09-10

  * fix(docs): update example import order
  * feat(upgrade): use gentleman v2

## 0.1.2 / 03-04-2016

- feat(consul): use Consul health services instead of from catalog.
- feat(api): due to Consul change, one field of the Consul struct interface has changed, but the struct is only used internally.
- refactor(examples): adapt retry example to use Consul health mock.

## 0.1.1 / 12-03-2016

- Maintenance release.

## 0.1.0 / 10-03-2016

- First release.


config_test.go
package consul

import (
	"testing"

	"github.com/nbio/st"
)

func TestNewConfig(t *testing.T) {
	config := NewConfig("server", "foo")
	st.Expect(t, config.Client.Address, "server")
	st.Expect(t, config.Service, "foo")
	st.Expect(t, config.Scheme, Scheme)
	st.Expect(t, config.Cache, true)
	st.Expect(t, config.Retry, true)
}

func TestNewBasicConfig(t *testing.T) {
	config := NewBasicConfig("server")
	st.Expect(t, config.Client.Address, "server")
	st.Expect(t, config.Service, "")
	st.Expect(t, config.Scheme, Scheme)
	st.Expect(t, config.Cache, true)
	st.Expect(t, config.Retry, true)
}


consul_test.go
package consul

import (
	"testing"
	"time"

	"github.com/nbio/st"
	"gopkg.in/eapache/go-resiliency.v1/retrier"
	"gopkg.in/h2non/gentleman-mock.v2"
	"gopkg.in/h2non/gentleman.v2"
	"gopkg.in/h2non/gock.v1"
)

const consulValidResponse = `
[
  {
    "Node":{
      "Node":"consul-client-nyc3-1",
      "Address":"127.0.0.1",
      "TaggedAddresses":{
        "wan":"127.0.0.1"
      },
      "CreateIndex":7,
      "ModifyIndex":375588
    },
    "Service":{
      "ID":"web",
      "Service":"web",
      "Tags":null,
      "Address":"",
      "Port":80,
      "EnableTagOverride":false,
      "CreateIndex":13,
      "ModifyIndex":13
    }
  }
]`

func TestConsulClient(t *testing.T) {
	defer gock.Off()

	config := NewConfig("demo.consul.io", "web")
	consul := New(config)
	gock.InterceptClient(config.Client.HttpClient)

	gock.New("http://demo.consul.io").
		Get("/v1/health/service/web").
		Reply(200).
		Type("json").
		BodyString(consulValidResponse)

	gock.New("http://127.0.0.1:80").
		Get("/").
		Reply(200).
		BodyString("hello world")

	cli := gentleman.New()
	cli.Use(mock.Plugin)
	cli.Use(consul)

	res, err := cli.Request().Send()
	st.Expect(t, err, nil)
	st.Expect(t, res.StatusCode, 200)
	st.Expect(t, res.String(), "hello world")
}

func TestConsulRetry(t *testing.T) {
	defer gock.Off()

	config := NewConfig("demo.consul.io", "web")
	consul := New(config)
	gock.InterceptClient(config.Client.HttpClient)

	gock.New("http://demo.consul.io").
		Get("/v1/health/service/web").
		Reply(200).
		Type("json").
		BodyString(consulValidResponse)

	gock.New("http://127.0.0.1:80").
		Get("/").
		Times(3).
		Reply(503)

	gock.New("http://127.0.0.1:80").
		Get("/").
		Reply(200).
		BodyString("hello world")

	cli := gentleman.New()
	cli.Use(mock.Plugin)
	cli.Use(consul)

	res, err := cli.Request().Send()
	st.Expect(t, err, nil)
	st.Expect(t, res.StatusCode, 200)
	st.Expect(t, res.String(), "hello world")
	st.Expect(t, gock.IsPending(), false)
}

func TestConsulRetryCustomStrategy(t *testing.T) {
	defer gock.Off()

	config := NewConfig("demo.consul.io", "web")
	config.Retrier = retrier.New(retrier.ConstantBackoff(10, time.Duration(25*time.Millisecond)), nil)
	consul := New(config)
	gock.InterceptClient(config.Client.HttpClient)

	gock.New("http://demo.consul.io").
		Get("/v1/health/service/web").
		Reply(200).
		Type("json").
		BodyString(consulValidResponse)

	gock.New("http://127.0.0.1:80").
		Get("/").
		Times(9).
		Reply(503)

	gock.New("http://127.0.0.1:80").
		Get("/").
		Reply(200).
		BodyString("hello world")

	cli := gentleman.New()
	cli.Use(mock.Plugin)
	cli.Use(consul)

	res, err := cli.Request().Send()
	st.Expect(t, err, nil)
	st.Expect(t, res.StatusCode, 200)
	st.Expect(t, res.String(), "hello world")
	st.Expect(t, gock.IsPending(), false)
}

func TestConsulDisableCache(t *testing.T) {
	defer gock.Off()

	config := NewConfig("demo.consul.io", "web")
	config.Cache = false
	consul := New(config)
	gock.InterceptClient(config.Client.HttpClient)

	gock.New("http://demo.consul.io").
		Get("/v1/health/service/web").
		Times(4).
		Reply(200).
		Type("json").
		BodyString(consulValidResponse)

	gock.New("http://127.0.0.1:80").
		Get("/").
		Times(3).
		Reply(503)

	gock.New("http://127.0.0.1:80").
		Get("/").
		Reply(200).
		BodyString("hello world")

	cli := gentleman.New()
	cli.Use(mock.Plugin)
	cli.Use(consul)

	res, err := cli.Request().Send()
	st.Expect(t, err, nil)
	st.Expect(t, res.StatusCode, 200)
	st.Expect(t, res.String(), "hello world")
	st.Expect(t, gock.IsPending(), false)
}

func TestConsulRetryDisabled(t *testing.T) {
	defer gock.Off()

	config := NewConfig("demo.consul.io", "web")
	config.Retry = false
	consul := New(config)
	gock.InterceptClient(config.Client.HttpClient)

	gock.New("http://demo.consul.io").
		Get("/v1/health/service/web").
		Reply(200).
		Type("json").
		BodyString(consulValidResponse)

	gock.New("http://127.0.0.1:80").
		Get("/").
		Reply(503)

	cli := gentleman.New()
	cli.Use(mock.Plugin)
	cli.Use(consul)

	res, err := cli.Request().Send()
	st.Expect(t, err, nil)
	st.Expect(t, res.StatusCode, 503)
	st.Expect(t, res.String(), "")
	st.Expect(t, gock.IsPending(), false)
}

const consulMultipleServers = `
[
  {
    "Node":{
      "Node":"consul-client-nyc3-1",
      "Address":"127.0.0.1",
      "TaggedAddresses":{
        "wan":"127.0.0.1"
      },
      "CreateIndex":7,
      "ModifyIndex":375588
    },
    "Service":{
      "ID":"web",
      "Service":"web",
      "Tags":null,
      "Address":"",
      "Port":80,
      "EnableTagOverride":false,
      "CreateIndex":13,
      "ModifyIndex":13
    }
  },
  {
    "Node":{
      "Node":"consul-client-nyc3-1",
      "Address":"127.0.0.2",
      "TaggedAddresses":{
        "wan":"127.0.0.2"
      },
      "CreateIndex":7,
      "ModifyIndex":375588
    },
    "Service":{
      "ID":"web",
      "Service":"web",
      "Tags":null,
      "Address":"",
      "Port":80,
      "EnableTagOverride":false,
      "CreateIndex":13,
      "ModifyIndex":13
    }
  },
  {
    "Node":{
      "Node":"consul-client-nyc3-1",
      "Address":"127.0.0.3",
      "TaggedAddresses":{
        "wan":"127.0.0.3"
      },
      "CreateIndex":7,
      "ModifyIndex":375588
    },
    "Service":{
      "ID":"web",
      "Service":"web",
      "Tags":null,
      "Address":"",
      "Port":80,
      "EnableTagOverride":false,
      "CreateIndex":13,
      "ModifyIndex":13
    }
  }
]`

func TestConsulNextServerFallback(t *testing.T) {
	defer gock.Off()

	config := NewConfig("demo.consul.io", "web")
	consul := New(config)
	gock.InterceptClient(config.Client.HttpClient)

	gock.New("http://demo.consul.io").
		Get("/v1/health/service/web").
		Reply(200).
		Type("json").
		BodyString(consulMultipleServers)

	gock.New("http://127.0.0.1:80").
		Get("/").
		Reply(503)

	gock.New("http://127.0.0.2:80").
		Get("/").
		Reply(503)

	gock.New("http://127.0.0.3:80").
		Get("/").
		Reply(200).
		BodyString("hello world")

	cli := gentleman.New()
	cli.Use(mock.Plugin)
	cli.Use(consul)

	res, err := cli.Request().Send()
	st.Expect(t, err, nil)
	st.Expect(t, res.StatusCode, 200)
	st.Expect(t, res.String(), "hello world")
	st.Expect(t, gock.IsPending(), false)
}


retrier_test.go
package consul

import (
	"testing"

	"github.com/nbio/st"
	"gopkg.in/h2non/gentleman.v2/context"
)

func TestRetrier(t *testing.T) {
	consul := NewClient(NewConfig("consul.server", "foo"))
	retrier := &Retrier{Consul: consul, Context: context.New(), Retry: DefaultRetrier}

	calls := 0
	retrier.Run(func() error {
		calls++
		return nil
	})

	st.Expect(t, calls, 1)
}

func TestNewRetrier(t *testing.T) {
	consul := NewClient(NewConfig("consul.server", "foo"))
	retrier := NewRetrier(consul, context.New())

	calls := 0
	retrier.Run(func() error {
		calls++
		return nil
	})

	st.Expect(t, calls, 1)
}


consul.go
package consul

import (
	"errors"
	"fmt"
	"sync"
	"time"

	"github.com/hashicorp/consul/api"
	"gopkg.in/h2non/gentleman-retry.v2"
	"gopkg.in/h2non/gentleman.v2/context"
	"gopkg.in/h2non/gentleman.v2/plugin"
)

// Consul represents the Consul plugin adapter for gentleman,
// which encapsulates the official Consul client and plugin specific settings.
type Consul struct {
	// mutex is used internallt to avoid race coditions for multithread scenarios.
	sync.Mutex

	// updated stores the last Consul update date.
	updated time.Time

	// cache is used to store and cache the servers.
	cache []string

	// Config stores the Consul's plugin specific settings.
	Config *Config

	// Client stores the official Consul client.
	Client *api.Client
}

// New creates a new Consul client with the given config
// and returns the gentleman plugin.
func New(config *Config) plugin.Plugin {
	return NewClient(config).Plugin()
}

// NewClient creates a new Consul high-level client.
func NewClient(config *Config) *Consul {
	client, _ := api.NewClient(config.Client)
	return &Consul{Config: config, Client: client}
}

// Plugin returns the gentleman plugin to be plugged.
func (c *Consul) Plugin() plugin.Plugin {
	handlers := plugin.Handlers{"before dial": c.OnBeforeDial}
	return &plugin.Layer{Handlers: handlers}
}

// IsUpdated returns true if the current list of catalog services is up-to-date,
// based on the cache TTL.
func (c *Consul) IsUpdated() bool {
	return len(c.cache) > 0 && time.Duration((time.Now().UnixNano()-c.updated.UnixNano())) < c.Config.CacheTTL
}

// UpdateCache updates the list of catalog services.
func (c *Consul) UpdateCache(nodes []string) {
	if !c.Config.Cache || len(nodes) == 0 {
		return
	}

	c.updated = time.Now()
	c.cache = nodes
}

// GetNodes returns a list of nodes for the current service from Consul server
// or from cache (if enabled and not expired).
func (c *Consul) GetNodes() ([]string, error) {
	c.Lock()
	defer c.Unlock()

	if c.IsUpdated() {
		return c.cache, nil
	}

	entries, _, err := c.Client.Health().Service(c.Config.Service, c.Config.Tag, true, c.Config.Query)
	if err != nil {
		return nil, err
	}

	nodes := makeInstances(entries)
	c.UpdateCache(nodes)

	return nodes, nil
}

func makeInstances(entries []*api.ServiceEntry) []string {
	instances := make([]string, len(entries))

	for i, entry := range entries {
		addr := entry.Node.Address
		if entry.Service.Address != "" {
			addr = entry.Service.Address
		}
		instances[i] = fmt.Sprintf("%s:%d", addr, entry.Service.Port)
	}

	return instances
}

// SetServerURL sets the request URL fields based on the given Consul service instance.
func (c *Consul) SetServerURL(ctx *context.Context, host string) {
	// Define server URL based on the best node
	ctx.Request.URL.Scheme = c.Config.Scheme
	ctx.Request.URL.Host = host
}

// GetBestCandidateNode retrieves and returns the best service node candidate
// asking to Consul server catalog or reading catalog from cache.
func (c *Consul) GetBestCandidateNode(ctx *context.Context) (string, error) {
	nodes, err := c.GetNodes()
	if err != nil {
		return "", err
	}
	if len(nodes) == 0 {
		return "", errors.New("consul: missing servers for service: " + c.Config.Service)
	}

	index := 0
	if retries, ok := ctx.Get("$consul.retries").(int); ok {
		index = retries
	}

	if index < len(nodes) {
		return nodes[index], nil
	}

	return nodes[0], nil
}

// UseBestCandidateNode sets the best service node URL in the given gentleman context.
func (c *Consul) UseBestCandidateNode(ctx *context.Context) error {
	node, err := c.GetBestCandidateNode(ctx)
	if err != nil {
		return err
	}

	// Define the proper URL in the outgoing request
	c.SetServerURL(ctx, node)
	return nil
}

// OnBeforeDial is a middleware function handler that replaces
// the outgoing request URL and provides a new http.RoundTripper if necessary
// in order to handle request failures and retry it accordingly.
func (c *Consul) OnBeforeDial(ctx *context.Context, h context.Handler) {
	// Define the server retries
	ctx.Set("$consul.retries", 0)

	// Use best service node candidate
	err := c.UseBestCandidateNode(ctx)
	if err != nil {
		h.Error(ctx, err)
		return
	}

	// Always continue with the next middleware
	defer h.Next(ctx)

	// If retry is disable, just continue
	if !c.Config.Retry {
		return
	}

	// Wrap HTTP transport with Consul retrier, if enabled
	retrier := NewRetrier(c, ctx)
	if c.Config.Retrier != nil {
		retrier.Retry = c.Config.Retrier
	}
	retry.InterceptTransport(ctx, retrier)
}


retrier.go
package consul

import (
	"gopkg.in/h2non/gentleman-retry.v2"
	"gopkg.in/h2non/gentleman.v2/context"
)

// DefaultRetrier stores the default retry strategy used by the plugin.
// By default will use a constant retry strategy with a maximum of 3 retry attempts.
var DefaultRetrier = retry.ConstantBackoff

// Retrier provides a retry.Retrier capable interface that
// encapsulates Consul client and user defined strategy.
type Retrier struct {
	// Consul stores the Consul client wrapper instance.
	Consul *Consul

	// Context stores the HTTP current gentleman context.
	Context *context.Context

	// Retry stores the retry strategy to be used.
	Retry retry.Retrier
}

// NewRetrier creates a default retrier for the given Consul client and context.
func NewRetrier(c *Consul, ctx *context.Context) *Retrier {
	return &Retrier{Consul: c, Context: ctx, Retry: DefaultRetrier}
}

// Run runs the given function multiple times, acting like a proxy
// to user defined retry strategy.
func (r *Retrier) Run(fn func() error) error {
	return r.Retry.Run(func() error {
		retries := 0
		if val, ok := r.Context.Get("$consul.retries").(int); ok {
			retries = val
		}

		// Expose number of retries
		defer r.Context.Set("$consul.retries", retries+1)

		// Call the function directly for the first attempt
		if retries == 0 {
			return fn()
		}

		// Set best server candidate
		err := r.Consul.UseBestCandidateNode(r.Context)
		if err != nil {
			return err
		}

		return fn()
	})
}


_examples/config/config.go
package main

import (
	"fmt"

	"gopkg.in/h2non/gentleman-consul.v2"
	"gopkg.in/h2non/gentleman.v2"
)

func main() {
	// Create a new client
	cli := gentleman.New()

	// Define custom Consul config settings
	config := consul.NewBasicConfig("demo.consul.io")
	config.Service = "web"
	config.Cache = false
	config.Retry = true

	// Register Consul's plugin at client level
	cli.Use(consul.New(config))

	// Create a new request based on the current client
	req := cli.Request()

	// Set a new header field
	req.SetHeader("Client", "gentleman")

	// Perform the request
	res, err := req.Send()
	if err != nil {
		fmt.Printf("Request error: %s\n", err)
		return
	}
	if !res.Ok {
		fmt.Printf("Invalid server response: %d\n", res.StatusCode)
		return
	}

	// Print response info
	fmt.Printf("Server URL: %s\n", res.RawRequest.URL.String())
	fmt.Printf("Response status: %d\n", res.StatusCode)
	fmt.Printf("Server header: %s\n", res.Header.Get("Server"))
}


_examples/discovery/discovery.go
package main

import (
	"fmt"

	"gopkg.in/h2non/gentleman-consul.v2"
	"gopkg.in/h2non/gentleman.v2"
)

func main() {
	// Create a new client
	cli := gentleman.New()

	// Register Consul's plugin at client level
	cli.Use(consul.New(consul.NewConfig("demo.consul.io", "web")))

	// Create a new request based on the current client
	req := cli.Request()

	// Set a new header field
	req.SetHeader("Client", "gentleman")

	// Perform the request
	res, err := req.Send()
	if err != nil {
		fmt.Printf("Request error: %s\n", err)
		return
	}
	if !res.Ok {
		fmt.Printf("Invalid server response: %d\n", res.StatusCode)
		return
	}

	// Print response info
	fmt.Printf("Server URL: %s\n", res.RawRequest.URL.String())
	fmt.Printf("Response status: %d\n", res.StatusCode)
	fmt.Printf("Server header: %s\n", res.Header.Get("Server"))
}


_examples/retry_strategy/retry.go
package main

import (
	"fmt"
	"time"

	"gopkg.in/eapache/go-resiliency.v1/retrier"
	"gopkg.in/h2non/gentleman-consul.v2"
	"gopkg.in/h2non/gentleman-mock.v2"
	"gopkg.in/h2non/gentleman.v2"
	"gopkg.in/h2non/gock.v1"
)

const consulValidResponse = `
[
  {
    "Node":{
      "Node":"consul-client-nyc3-1",
      "Address":"127.0.0.1",
      "TaggedAddresses":{
        "wan":"127.0.0.1"
      },
      "CreateIndex":7,
      "ModifyIndex":375588
    },
    "Service":{
      "ID":"web",
      "Service":"web",
      "Tags":null,
      "Address":"",
      "Port":80,
      "EnableTagOverride":false,
      "CreateIndex":13,
      "ModifyIndex":13
    }
  }
]`

func main() {
	defer gock.Off()

	// Mock consul server
	gock.New("http://demo.consul.io").
		Get("/v1/health/service/web").
		Reply(200).
		Type("json").
		BodyString(consulValidResponse)

	// Configure failure responses
	gock.New("http://127.0.0.1:80").
		Get("/").
		Times(10).
		Reply(503)

	// Final valid response
	gock.New("http://127.0.0.1:80").
		Get("/").
		Reply(200).
		SetHeader("Server", "gock").
		BodyString("hello world")

	// Create a new client
	cli := gentleman.New()

	// Configure Consul plugin
	config := consul.NewConfig("demo.consul.io", "web")

	// Use a custom retrier strategy with max 10 retry attempts
	config.Retrier = retrier.New(retrier.ConstantBackoff(10, time.Duration(25*time.Millisecond)), nil)

	// Intercept HTTP transport via gock to simulate the failures
	gock.InterceptClient(config.Client.HttpClient)
	cli.Use(mock.Plugin)

	// Register Consul's plugin at client level
	cli.Use(consul.New(config))

	// Create a new request based on the current client
	req := cli.Request()

	// Set a new header field
	req.SetHeader("Client", "gentleman")

	// Perform the request
	res, err := req.Send()
	if err != nil {
		fmt.Printf("Request error: %s\n", err)
		return
	}
	if !res.Ok {
		fmt.Printf("Invalid server response: %d\n", res.StatusCode)
		return
	}

	// Print response info
	fmt.Printf("Server URL: %s\n", res.RawRequest.URL.String())
	fmt.Printf("Response status: %d\n", res.StatusCode)
	fmt.Printf("Server header: %s\n", res.Header.Get("Server"))
}
